File: clients/ios/VisualPose/OverlayMath.swift
import Foundation

struct CocoKeypoint {
    var x: Float
    var y: Float
    var confidence: Float
}

struct CocoPose {
    var keypoints: [CocoKeypoint]
}

struct OverlayCover {
    var scale: Float
    var dx: Float
    var dy: Float
}

struct OverlaySegment {
    var x1: Float
    var y1: Float
    var x2: Float
    var y2: Float
}

struct OverlayJoint {
    var x: Float
    var y: Float
}

enum OverlayMath {
    static let minVis: Float = 0.3
    static let coco17Edges: [(Int, Int)] = [
        (0, 1), (0, 2), (1, 3), (2, 4), (5, 6), (5, 7), (7, 9), (6, 8), (8, 10),
        (5, 11), (6, 12), (11, 12), (11, 13), (13, 15), (12, 14), (14, 16),
    ]

    static func cover(srcW: Int, srcH: Int, viewW: Int, viewH: Int) -> OverlayCover {
        guard srcW > 0, srcH > 0, viewW > 0, viewH > 0 else {
            return OverlayCover(scale: 1, dx: 0, dy: 0)
        }
        let scale = max(Float(viewW) / Float(srcW), Float(viewH) / Float(srcH))
        return OverlayCover(
            scale: scale,
            dx: (Float(viewW) - Float(srcW) * scale) / 2,
            dy: (Float(viewH) - Float(srcH) * scale) / 2
        )
    }

    static func contain(srcW: Int, srcH: Int, viewW: Int, viewH: Int) -> OverlayCover {
        guard srcW > 0, srcH > 0, viewW > 0, viewH > 0 else {
            return OverlayCover(scale: 1, dx: 0, dy: 0)
        }
        let scale = min(Float(viewW) / Float(srcW), Float(viewH) / Float(srcH))
        return OverlayCover(
            scale: scale,
            dx: (Float(viewW) - Float(srcW) * scale) / 2,
            dy: (Float(viewH) - Float(srcH) * scale) / 2
        )
    }

    static func mapContain(x: Float, y: Float, srcW: Int, srcH: Int, viewW: Int, viewH: Int) -> (Float, Float) {
        let box = contain(srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH)
        return (x * box.scale + box.dx, y * box.scale + box.dy)
    }

    static func unmapContain(x: Float, y: Float, srcW: Int, srcH: Int, viewW: Int, viewH: Int) -> (Float, Float) {
        let box = contain(srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH)
        return ((x - box.dx) / box.scale, (y - box.dy) / box.scale)
    }

    static func mapFillCenter(x: Float, y: Float, srcW: Int, srcH: Int, viewW: Int, viewH: Int) -> (Float, Float) {
        let box = cover(srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH)
        return (x * box.scale + box.dx, y * box.scale + box.dy)
    }

    static func mapPoint(
        x: Float, y: Float, srcW: Int, srcH: Int, viewW: Int, viewH: Int, letterbox: Bool
    ) -> (Float, Float) {
        if letterbox {
            return mapContain(x: x, y: y, srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH)
        }
        return mapFillCenter(x: x, y: y, srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH)
    }

    static func visibleJoints(
        pose: CocoPose, srcW: Int, srcH: Int, viewW: Int, viewH: Int, letterbox: Bool = true
    ) -> [OverlayJoint] {
        pose.keypoints.compactMap { kp in
            guard kp.confidence >= minVis else { return nil }
            let (x, y) = mapPoint(
                x: kp.x, y: kp.y, srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH, letterbox: letterbox
            )
            return OverlayJoint(x: x, y: y)
        }
    }

    static func visibleSegments(
        pose: CocoPose, srcW: Int, srcH: Int, viewW: Int, viewH: Int, letterbox: Bool = true
    ) -> [OverlaySegment] {
        guard pose.keypoints.count >= 17 else { return [] }
        return coco17Edges.compactMap { a, b in
            let pa = pose.keypoints[a]
            let pb = pose.keypoints[b]
            guard pa.confidence >= minVis, pb.confidence >= minVis else { return nil }
            let (x1, y1) = mapPoint(
                x: pa.x, y: pa.y, srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH, letterbox: letterbox
            )
            let (x2, y2) = mapPoint(
                x: pb.x, y: pb.y, srcW: srcW, srcH: srcH, viewW: viewW, viewH: viewH, letterbox: letterbox
            )
            return OverlaySegment(x1: x1, y1: y1, x2: x2, y2: y2)
        }
    }

    static func bboxFromPose(_ pose: CocoPose) -> BBox? {
        let pts = pose.keypoints.filter { $0.confidence >= minVis }
        guard !pts.isEmpty else { return nil }
        return BBox(
            x1: pts.map(\.x).min()!,
            y1: pts.map(\.y).min()!,
            x2: pts.map(\.x).max()!,
            y2: pts.map(\.y).max()!
        )
    }
}

File: clients/android/app/src/main/java/local/visual/corepose/OverlayMath.kt
package local.visual.corepose

data class CocoKeypoint(
    val x: Float,
    val y: Float,
    val confidence: Float,
)

data class CocoPose(
    val keypoints: List<CocoKeypoint>,
)

object OverlayMath {
    const val MIN_VIS = 0.3f

    val COCO17_EDGES: List<Pair<Int, Int>> = listOf(
        0 to 1,
        0 to 2,
        1 to 3,
        2 to 4,
        5 to 6,
        5 to 7,
        7 to 9,
        6 to 8,
        8 to 10,
        5 to 11,
        6 to 12,
        11 to 12,
        11 to 13,
        13 to 15,
        12 to 14,
        14 to 16,
    )

    data class Segment(
        val x1: Float,
        val y1: Float,
        val x2: Float,
        val y2: Float,
    )

    data class Joint(
        val x: Float,
        val y: Float,
    )

    data class Cover(
        val scale: Float,
        val dx: Float,
        val dy: Float,
    )

    fun cover(srcW: Int, srcH: Int, viewW: Int, viewH: Int): Cover {
        if (srcW <= 0 || srcH <= 0 || viewW <= 0 || viewH <= 0) {
            return Cover(1f, 0f, 0f)
        }
        val scale = maxOf(viewW.toFloat() / srcW, viewH.toFloat() / srcH)
        val dx = (viewW - srcW * scale) / 2f
        val dy = (viewH - srcH * scale) / 2f
        return Cover(scale, dx, dy)
    }

    fun contain(srcW: Int, srcH: Int, viewW: Int, viewH: Int): Cover {
        if (srcW <= 0 || srcH <= 0 || viewW <= 0 || viewH <= 0) {
            return Cover(1f, 0f, 0f)
        }
        val scale = minOf(viewW.toFloat() / srcW, viewH.toFloat() / srcH)
        val dx = (viewW - srcW * scale) / 2f
        val dy = (viewH - srcH * scale) / 2f
        return Cover(scale, dx, dy)
    }

    fun mapFillCenter(
        x: Float,
        y: Float,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
    ): Pair<Float, Float> {
        val cover = cover(srcW, srcH, viewW, viewH)
        return x * cover.scale + cover.dx to y * cover.scale + cover.dy
    }

    fun unmapFillCenter(
        x: Float,
        y: Float,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
    ): Pair<Float, Float> {
        val cover = cover(srcW, srcH, viewW, viewH)
        return (x - cover.dx) / cover.scale to (y - cover.dy) / cover.scale
    }

    fun mapContain(
        x: Float,
        y: Float,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
    ): Pair<Float, Float> {
        val box = contain(srcW, srcH, viewW, viewH)
        return x * box.scale + box.dx to y * box.scale + box.dy
    }

    fun unmapContain(
        x: Float,
        y: Float,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
    ): Pair<Float, Float> {
        val box = contain(srcW, srcH, viewW, viewH)
        return (x - box.dx) / box.scale to (y - box.dy) / box.scale
    }

    fun mapPoint(
        x: Float,
        y: Float,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
        letterbox: Boolean,
    ): Pair<Float, Float> {
        return if (letterbox) {
            mapContain(x, y, srcW, srcH, viewW, viewH)
        } else {
            mapFillCenter(x, y, srcW, srcH, viewW, viewH)
        }
    }

    fun visibleJoints(
        pose: CocoPose,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
        letterbox: Boolean = false,
    ): List<Joint> {
        return pose.keypoints.mapNotNull { kp ->
            if (kp.confidence < MIN_VIS) {
                null
            } else {
                val (x, y) = mapPoint(kp.x, kp.y, srcW, srcH, viewW, viewH, letterbox)
                Joint(x, y)
            }
        }
    }

    fun visibleSegments(
        pose: CocoPose,
        srcW: Int,
        srcH: Int,
        viewW: Int,
        viewH: Int,
        letterbox: Boolean = false,
    ): List<Segment> {
        if (pose.keypoints.size < 17) {
            return emptyList()
        }
        return COCO17_EDGES.mapNotNull { (a, b) ->
            val pa = pose.keypoints[a]
            val pb = pose.keypoints[b]
            if (pa.confidence < MIN_VIS || pb.confidence < MIN_VIS) {
                null
            } else {
                val (x1, y1) = mapPoint(pa.x, pa.y, srcW, srcH, viewW, viewH, letterbox)
                val (x2, y2) = mapPoint(pb.x, pb.y, srcW, srcH, viewW, viewH, letterbox)
                Segment(x1, y1, x2, y2)
            }
        }
    }

    fun bboxFromPose(pose: CocoPose): BBox? {
        val pts = pose.keypoints.filter { it.confidence >= MIN_VIS }
        if (pts.isEmpty()) {
            return null
        }
        return BBox(
            pts.minOf { it.x },
            pts.minOf { it.y },
            pts.maxOf { it.x },
            pts.maxOf { it.y },
        )
    }
}
